### DISCUSSION TITLE
Error after the most recent commits for TBIW

### DISCUSSION THREAD

**Discussion by freznel10** [joshreve/dactyl-keyboard#28] [General]
Traceback (most recent call last):

File "/app/src/generate_configuration.py", line 365, in <module>

run()

File "/app/src/dactyl_manuform.py", line 3625, in run

mod_r = model_side(side="right")

File "/app/src/dactyl_manuform.py", line 3519, in model_side

shape = union([shape, sensor])

File "/app/src/helpers_cadquery.py", line 56, in union

shape = shape.union(item)

File "/opt/conda/lib/python3.9/site-packages/cadquery/cq.py", line 3242, in union

r = r.clean()

File "/opt/conda/lib/python3.9/site-packages/cadquery/occ_impl/shapes.py", line 389, in clean

upgrader.Build()

OCP.Standard.Standard_Failure: Courbes non jointives

>>>

**Comment by joshreve** [comment]
Can you send your config or json file?  Mine appears to be working.  I will also commit one more time just in case.  The engine has issues sometimes when coplanar objects are unioned.  Not sure if that is the issue or not, but I should be able to figure it out with your config.

**Comment by freznel10** [comment]
I pulled the latest commit and get this error:

Traceback (most recent call last):

File "/app/src/generate_configuration.py", line 374, in <module>

run()

File "/app/src/dactyl_manuform.py", line 3651, in run

mod_r = model_side(side="right")

File "/app/src/dactyl_manuform.py", line 3545, in model_side

shape = union([shape, sensor])

File "/app/src/helpers_cadquery.py", line 56, in union

shape = shape.union(item)

File "/opt/conda/lib/python3.9/site-packages/cadquery/cq.py", line 3242, in
union

r = r.clean()

File "/opt/conda/lib/python3.9/site-packages/cadquery/occ_impl/shapes.py",
line 389, in clean

upgrader.Build()

OCP.Standard.Standard_Failure: Courbes non jointives

>>>

Is this a module/dependency problem? I'm building using Docker containers.
Attaching my json here.

When I change the engine to solid, it builds. However, the generated file
fails to render in OpenSCAD.

Thanks

**Reply by joshreve** [reply]
I can't find the attachment.  I'm guessing there is something in the config that creates an odd geometry, but I will check to see if I can make things more robust.  Can you try building with the current default config that is checked in?  I have that running correctly and it will tell me if the issue is with the specific configuration or with how you configuration is operating.  I just ran through docker and had to delete and recreate the containers.  Not sure why they were misbehaving, but it ran after that.  

Either way, a quick check of your config will tell me the issue.  You can also just paste it into a reply since I don't think the attachment shows up.  Or if you have it in a forked repo, I'm happy to pull it from there as well.

**Comment by freznel10** [comment]
Here's my config:

{
    "ENGINE": "cadquery",
    "save_dir": ".",
    "config_name": "DM",
    "show_caps": false,
    "show_pcbs": false,
    "nrows": 5,
    "ncols": 6,
    "alpha": 0.26179916666666664,
    "beta": 0.08726638888888888,
    "centercol": 3,
    "centerrow_offset": 3,
    "tenting_angle": 0.26179916666666664,
    "symmetry": "symmetric",
    "column_style_gt5": "orthographic",
    "column_style": "standard",
    "thumb_offsets": [
        6,
        -3,
        7
    ],
    "keyboard_z_offset": 11,
    "thumb_style": "CARBONFET",
    "default_1U_cluster": false,
    "minidox_Usize": 1.6,
    "thumb_plate_tr_rotation": 0.0,
    "thumb_plate_tl_rotation": 0.0,
    "thumb_plate_mr_rotation": 0.0,
    "thumb_plate_ml_rotation": 0.0,
    "thumb_plate_br_rotation": 0.0,
    "thumb_plate_bl_rotation": 0.0,
    "trackball_in_wall": true,
    "tbiw_hole_diameter": 36,
    "tbiw_hole_height": 40,
    "tbiw_ball_center_row": 0.2,
    "tbiw_translational_offset": [
        0.0,
        0.0,
        3
    ],
    "tbiw_rotation_offset": [
        0.0,
        0.0,
        0.0
    ],
    "tbiw_left_wall_x_offset_override": 50.0,
    "tbiw_left_wall_z_offset_override": 0.0,
    "tbiw_left_wall_lower_y_offset": 0.0,
    "tbiw_left_wall_lower_z_offset": 0.0,
    "tbiw_oled_center_row": 0.75,
    "tbiw_oled_translation_offset": [
        -3.5,
        0,
        1.5
    ],
    "tbiw_oled_rotation_offset": [
        0,
        0,
        0
    ],
    "other_thumb": "DEFAULT",
    "tbjs_key_diameter": 64,
    "tbjs_translational_offsets": [
        [
            0.0,
            0.0,
            -3.0
        ],
        [
            0.0,
            0.0,
            -3.0
        ],
        [
            0.0,
            0.0,
            -3.0
        ],
        [
            0.0,
            0.0,
            -3.0
        ]
    ],
    "tbjs_rotation_offsets": [
        [
            0.0,
            0.0,
            0.0
        ],
        [
            0.0,
            0.0,
            0.0
        ],
        [
            0.0,
            0.0,
            0.0
        ],
        [
            0.0,
            0.0,
            0.0
        ]
    ],
    "trackball_Usize": 1.5,
    "trackball_rotation": 0.0,
    "ball_side": "right",
    "ball_diameter": 34.0,
    "ball_wall_thickness": 5.0,
    "ball_gap": 1.0,
    "ball_z_offset": -3.0,
    "pinky_1_5U": false,
    "first_1_5U_row": 0,
    "last_1_5U_row": 5,
    "extra_width": 2.5,
    "extra_height": 1.0,
    "wall_z_offset": 15,
    "wall_x_offset": 5,
    "wall_y_offset": 6,
    "left_wall_x_offset": 12,
    "left_wall_z_offset": 3,
    "left_wall_lower_y_offset": 0,
    "left_wall_lower_z_offset": 0,
    "wall_thickness": 4.5,
    "wall_base_y_thickness": 4.5,
    "wall_base_x_thickness": 4.5,
    "wall_base_back_thickness": 4.5,
    "fixed_angles": [
        0.17453277777777776,
        0.17453277777777776,
        0,
        0,
        0,
        -0.26179916666666664,
        -0.26179916666666664
    ],
    "fixed_x": [
        -41.5,
        -22.5,
        0,
        20.3,
        41.4,
        65.5,
        89.6
    ],
    "fixed_z": [
        12.1,
        8.3,
        0,
        5,
        10.7,
        14.5,
        17.5
    ],
    "fixed_tenting": 0.0,
    "plate_style": "NOTCH",
    "hole_keyswitch_height": 14.0,
    "hole_keyswitch_width": 14.0,
    "nub_keyswitch_height": 14.4,
    "nub_keyswitch_width": 14.4,
    "undercut_keyswitch_height": 14.0,
    "undercut_keyswitch_width": 14.0,
    "notch_width": 5.0,
    "sa_profile_key_height": 12.7,
    "sa_length": 18.5,
    "sa_double_length": 37.5,
    "plate_thickness": 5.1,
    "plate_rim": 2.0,
    "clip_thickness": 1.4,
    "clip_undercut": 1.0,
    "undercut_transition": 0.2,
    "plate_file": null,
    "plate_offset": 0.0,
    "oled_mount_type": "CLIP",
    "oled_center_row": 1.25,
    "oled_translation_offset": [
        0,
        0,
        4
    ],
    "oled_rotation_offset": [
        0,
        0,
        0
    ],
    "oled_configurations": {
        "UNDERCUT": {
            "oled_mount_width": 15.0,
            "oled_mount_height": 35.0,
            "oled_mount_rim": 3.0,
            "oled_mount_depth": 6.0,
            "oled_mount_cut_depth": 20.0,
            "oled_mount_location_xyz": [
                -80.0,
                20.0,
                45.0
            ],
            "oled_mount_rotation_xyz": [
                13.0,
                0.0,
                -6.0
            ],
            "oled_left_wall_x_offset_override": 28.0,
            "oled_left_wall_z_offset_override": 0.0,
            "oled_left_wall_lower_y_offset": 12.0,
            "oled_left_wall_lower_z_offset": 5.0,
            "oled_mount_undercut": 1.0,
            "oled_mount_undercut_thickness": 2.0
        },
        "SLIDING": {
            "oled_mount_width": 12.5,
            "oled_mount_height": 25.0,
            "oled_mount_rim": 2.5,
            "oled_mount_depth": 8.0,
            "oled_mount_cut_depth": 20.0,
            "oled_mount_location_xyz": [
                -78.0,
                10.0,
                41.0
            ],
            "oled_mount_rotation_xyz": [
                6.0,
                0.0,
                -3.0
            ],
            "oled_left_wall_x_offset_override": 24.0,
            "oled_left_wall_z_offset_override": 0.0,
            "oled_left_wall_lower_y_offset": 12.0,
            "oled_left_wall_lower_z_offset": 5.0,
            "oled_thickness": 4.2,
            "oled_edge_overlap_end": 6.5,
            "oled_edge_overlap_connector": 5.5,
            "oled_edge_overlap_thickness": 2.5,
            "oled_edge_overlap_clearance": 2.5,
            "oled_edge_chamfer": 2.0
        },
        "CLIP": {
            "oled_mount_width": 12.5,
            "oled_mount_height": 39.0,
            "oled_mount_rim": 2.0,
            "oled_mount_depth": 7.0,
            "oled_mount_cut_depth": 20.0,
            "oled_mount_location_xyz": [
                -78.0,
                20.0,
                42.0
            ],
            "oled_mount_rotation_xyz": [
                12.0,
                0.0,
                -6.0
            ],
            "oled_left_wall_x_offset_override": 24.0,
            "oled_left_wall_z_offset_override": 0.0,
            "oled_left_wall_lower_y_offset": 12.0,
            "oled_left_wall_lower_z_offset": 5.0,
            "oled_thickness": 4.2,
            "oled_mount_bezel_thickness": 3.5,
            "oled_mount_bezel_chamfer": 2.0,
            "oled_mount_connector_hole": 6.0,
            "oled_screen_start_from_conn_end": 6.5,
            "oled_screen_length": 24.5,
            "oled_screen_width": 10.5,
            "oled_clip_thickness": 1.5,
            "oled_clip_width": 6.0,
            "oled_clip_overhang": 1.0,
            "oled_clip_extension": 5.0,
            "oled_clip_width_clearance": 0.5,
            "oled_clip_undercut": 0.5,
            "oled_clip_undercut_thickness": 2.5,
            "oled_clip_y_gap": 0.2,
            "oled_clip_z_gap": 0.2
        }
    },
    "web_thickness": 4.0,
    "post_size": 0.1,
    "post_adj": 0,
    "screws_offset": "INSIDE",
    "screw_insert_height": 3.8,
    "screw_insert_bottom_radius": 2.655,
    "screw_insert_top_radius": 2.55,
    "wire_post_height": 7,
    "wire_post_overhang": 3.5,
    "wire_post_diameter": 2.6,
    "controller_mount_type": "EXTERNAL",
    "external_holder_height": 12.5,
    "external_holder_width": 28.75,
    "external_holder_xoffset": -5.0,
    "external_holder_yoffset": -4.5,
    "screw_hole_diameter": 3,
    "base_thickness": 3.0,
    "base_offset": 3.0,
    "base_rim_thickness": 5.0,
    "screw_cbore_diameter": 4.0,
    "screw_cbore_depth": 2.0,
    "plate_holes": true,
    "plate_holes_xy_offset": [
        0.0,
        0.0
    ],
    "plate_holes_width": 14.3,
    "plate_holes_height": 14.3,
    "plate_holes_diameter": 1.7,
    "plate_holes_depth": 20.0,
    "pcb_width": 18.0,
    "pcb_height": 18.0,
    "pcb_thickness": 1.5,
    "pcb_hole_diameter": 2,
    "pcb_hole_pattern_width": 14.3,
    "pcb_hole_pattern_height": 14.3,
    "column_offsets": [
        [
            0,
            0,
            0
        ],
        [
            0,
            0,
            0
        ],
        [
            0,
            2.82,
            -4.5
        ],
        [
            0,
            0,
            0
        ],
        [
            0,
            -6,
            5
        ],
        [
            0,
            -6,
            5
        ],
        [
            0,
            -6,
            5
        ]
    ]
}

**Reply by joshreve** [reply]
So good and bad, it renders for me, so it is something in your environment.  If you are using Docker I would consider deleting your apps and rebuilding them.  I had an issue with the Docker building and clearing them out and rebuilding fixed it quickly.  If you are on a Linux host I don't have that currently setup but the same could work.  If you are running on Linux let me know and I can check the setup.  

I feel like it has to do with the imported files as it fails on the trackball components.  You could try building with the hotswap to see if you get the same issue to narrow it down to the import process.

**Reply by freznel10** [reply]
I deleted all docker containers/images and reset everything but I still get the same error for TBIW when using cadquery. When I change to solid, TBIW builds. However, it fails to render in openscad because supposedly the polygons are not closed.

**Reply by joshreve** [reply]
What is your OS, setup, etc, maybe that will help.  Are you running the master or trackball branch?

**Reply by joshreve** [reply]
That was a stupid question since only the trackball branch has the tbiw, but the other questions stand.  I'm just not sure how to troubleshoot since your file work for me.  If you are on windows then I am really confused since that is my current config.

Try the hotswap sockets without the trackball.  That may help me debug.  

I'm assuming you pulled latest. . . Just trying to think of next steps.

I will also try a fresh clone to see if I have something in my files that is not making it into the repo properly.

**Comment by freznel10** [comment]
Windows 10 fresh install. Docker, WSL2 backend. Building using docker.
Running the trackball branch. No problems in the master branch whatsoever.
Master branch works fine. Only running into errors in the trackball branch
specifically TBIW.

For the trackball branch:
Only things I changed in the config are CARBONFET thumb cluster and TBIW
set to true.

It builds in solid, previews fine in openSCAD, but render fails due to
unclosed polygons/mesh.

For CQ, it fails to build completely and I get the errors I posted before.

I did a fresh install on an new machine since installing miniconda and all
other dependencies somehow messes up ZMK. So I’m using a new machine write
a fresh install to do the tests.
Thank you!

**Comment by freznel10** [comment]
Alright, I will try the hot swap pockets and report back.

**Reply by joshreve** [reply]
If that fails, I'll have to think about it a bit more.  If the hot swap works you can try commenting out the actual trackball geometry in the code with the failing config, but I want to see what happens with hot swap first.  You could also try the trackball thumb that was just added.  I'm trying to see if it has to do with the trackball itself and the imports involved.  

I did a fresh clone with no issues, so it has to be with the setup. . . . One other thought, do you have odd characters in the path or an excessively long path?  It might be a linux vs windows naming issue.  If so, the hotswap test will fail as well.

**Reply by freznel10** [reply]
The hotswap build works. I tried the trackball thumb and it works as well. Ill try to transfer directories and see if its a naming issue.

**Reply by joshreve** [reply]
If those work, I am really not sure.  Try moving it around in XY with  'tbiw_translational_offset': (0.0, 0.0, 3).  Maybe you have some marginal condition.  The thumb uses the same parts and is "harder" to construct, so I am a bit at a loss.  Sometimes you can get weird conditions on relative positioning and tolerances when it goes to union and shifting it may break it out of that issue.

**Comment by freznel10** [comment]
I'll try that. I'm getting a new machine and install everything fresh. Ill
ty to screen record the entire process and send a link to the video.

**Comment by freznel10** [comment]
I just used the latest commit. OpenSCAD file opens fine but when ti try to
render I get this error:

Rendering Polygon Mesh using CGAL...

ERROR: The given mesh is not closed! Unable to convert to
CGAL_Nef_Polyhedron.

Geometries in cache: 1938

Geometry cache size in bytes: 15725800

CGAL Polyhedrons in cache: 4

CGAL cache size in bytes: 91245328

Total rendering time: 0:03:46.781

Top level object is a 3D object:

Simple: yes

Vertices: 13919

Halfedges: 62934

Edges: 31467

Halffacets: 35090

Facets: 17545

Volumes: 2

Rendering finished


The render has a hole where the sensor lens go. I opened the trackball
models in Blender and one of them has non-manifold geometry (one of the
sensor models).

Im using docker a new machine.

**Reply by nethermead** [reply]
I ran into this today for trackball builds.  The tracker_sensor_mount.stl has errors.  I can render in OpenSCAD (with similar errors to the above) but generated STLs are missing the sensor holder entirely.  CQ renders for me and I can get an STL out of it and I'm now trying another test print of the trackball cluster.  I've tried a couple of prints before from CQ but PrusaSlicer always goes wonky near the sensor hole.  The resulting prints come out ragged around the hole with a section of the holder apparently printed in mid-air. I did a couple of tweaks to the config and reoriented the cluster test to see if that helps.

**Reply by joshreve** [reply]
I think I need to go to a simpler geometry there.  I am trying to match the spec sheet, but it creates a sharp angle intersection with a revolve which seems to be a real issue when translated to stl.  I know this needs fixing.  I got the cadquery version to print without an issue.  Ideally I will translate this to geometric functions rather than an import and it should fix these issues.

I am aware of the issue, but completing a build at the moment to find the rest of the issues before doing a full round of cleanup.  If anyone is interested in recreating with the helper function selection that would be hugely useful as that will be my future intent once I get through the initial build.  

Check the dev build.  @freznel10 fixed the geometry so it should be better in that build.  I still want to get to directly generated vs imported, but that should fix the problem until that happens (which could be a while).
